numba/core/typing/builtins.py

```diff
@@ -940,6 +941,8 @@ def generic(self, args, kws):
         if isinstance(arg, (types.Float, types.Boolean)):
             return signature(types.intp, arg)
 
+        raise TypeError("int() only support for Integer, Float and Boolean")
+
```

Comment:
I suppose you just can't raise a TypeError here. This method (`generic`) is called from `AbtractTemplate::apply` (1) and there's a note there saying it is perfectly ok for `generic` to return either a signature or None. In the `Int` case above, if there isn't a match, `generic()` will return nothing (or None). 

(1) https://github.com/numba/numba/blob/3461bcb76a479bffbadbef24b6662107229e2a8a/numba/core/typing/templates.py#L343-L366

If the type you're calling is an `Optional<T>` type, `apply()` will first unwrap it and then try `generic()` again. I suppose it is safe to raise the TypeError only if the type is not an instance of Optional.

Reply:
Yeah, it was returning `None` before, raising an error causes the error message to be included in the final error/warning message. The idea here is that you might call `int()` with a valid signature by the python docs, but not supported by numba, the error message is not clear in that case.

For example, trying to convert a string to integer, which is allowed in python, says no implementation found for `int(unicode_type)`. In my case when I saw that my thought process was "There is probably some problem converting between unicode, str and bytes like back in Python 2", spent quite a lot of time on that.

This is nothing new, the same is already being done for `float()`, I copied the idea from there.

Reply:
Yes, but you have a failing test that calls `int()` with an optional type. I suppose the following is ok:
```suggestion
    if not isinstance(arg, types.Optional):
        raise TypeError("int() only support for Integer, Float and Boolean")
```

Reply:
Now I got what you meant by that. I can't raise it as a default, not that I can't raise it at all. Sorry